Fix #7242, CALL(@0) crash backend

c2main opened the pull request:
When executing a prepared CALL, which is not pure SQL but available with some drivers like npgsql and jpgdbc, Citus entered a code path where a plan is not defined, while trying to increase its cost. Thus SIG11 when plan is a NULL pointer.

Fix by only increasing plan cost when plan is not null.

However, it is a bit suspicious to get here with a NULL plan and maybe a better change will be to not call ShardPlacementForFunctionColocatedWithDistTable() with a NULL plan at all (in call.c:134)

bug hit with for example:
```
CallableStatement proc = con.prepareCall("{CALL p(?)}");
proc.registerOutParameter(1, java.sql.Types.BIGINT);
proc.setInt(1, -100);
proc.execute();
```

where `p(bigint)` is a distributed "function" and the param the distribution key (also in a distributed table), see #7242 for details

Fixes #7242

c2main commented:
Oh, and the question is: does citus provides a test framework for testing such corner case ?

JelteF commented on src/backend/distributed/planner/function_call_delegation.c:
```diff
@@ -526,7 +526,10 @@ ShardPlacementForFunctionColocatedWithDistTable(DistObjectCacheEntry *procedure,
 		{
 			/* Don't log a message, we should end up here again without a parameter */
-			DissuadePlannerFromUsingPlan(plan);
+			if (plan)
```
I think it's worth updating the comment that sometimes there is no plan at this stage at all yet, for instance when a `CALL` statement is prepared.

c2main replied:
yep, applied.

JelteF reviewed (approved):
Thanks for this. Change looks good to me now. (CI might still complain though)

JelteF commented:
FYI CI complains about some unused imports in the test file

c2main commented:
> FYI CI complains about some unused imports in the test file, if you fix those I think this is good to merge.

I used other tests as examples...
I removed the "imports" and rebased.

It will also probably be interesting to remove "imports" from [test_columnar.py](https://github.com/citusdata/citus/blob/main/src/test/regress/citus_tests/test/test_columnar.py) and [test_extension.py](https://github.com/citusdata/citus/blob/main/src/test/regress/citus_tests/test/test_extension.py)

c2main commented:
> > FYI CI complains about some unused imports in the test file, if you fix those I think this is good to merge.
> 
> I used other tests as examples... I removed the "imports" and rebased.
> 
> It will also probably be interesting to remove "imports" from [test_columnar.py](https://github.com/citusdata/citus/blob/main/src/test/regress/citus_tests/test/test_columnar.py) and [test_extension.py](https://github.com/citusdata/citus/blob/main/src/test/regress/citus_tests/test/test_extension.py)

Well, no. Those do really uses `pytest`
